### DISCUSSION TITLE
Script doesn't terminate itself when expected to

### DISCUSSION THREAD

**Discussion by EmilyGraceSeville7cf** [charmbracelet/gum#647] [Q&A]
Hi everyone! I have a task to write a script to create a terminal-based menu for
launching `flatpak` applications (referred later as **launcher**). **launcher**
should allow me to choose the application to run and then close itself when the
picked application is started.

To implement this I've done this:

- have created a `gui.fish` script (referred later as **gui**) to execute **launcher**
  and pass several additional options for `gnome-terminal` like `--zoom 1.5`
  (full code is available [here](https://github.com/EmilyGraceSeville7cf/desktop-environment/blob/main/gui.fish))
- have created a **launcher** script to launch an actual `flatpak` application
  (full code is available [here](https://github.com/EmilyGraceSeville7cf/desktop-environment/blob/main/open-flatpak-application.fish))

The problem is that the **gui** is not closed when the `flatpak` application is started.

To reproduce this issue I've found a minimal code to show what's going on:

**gui**:

```fish
#!/usr/bin/env fish

set menu_path $argv[1] # path to the launcher

gnome-terminal -- $menu_path # <-- doesn't close itself when expected to
```

and **launcher**:

```fish
#!/usr/bin/env fish

gum spin -- fish --command 'flatpak run org.mozilla.firefox &'
```

How to fix this issue?

Edit: partial solution is to use `gum spin -- setsid flatpak run org.mozilla.firefox` but it
just prevents chosen application from being closed when the terminal is terminated, it doesn't
force terminal to close itself. And from what I've noticed the terminal instance closes just
when it's not the first terminal instance.

**Comment by wellkilo** [comment]
The terminal remains open because the launched Flatpak process still inherits file descriptors from the command that gum spin is supervising. Backgrounding with & or calling setsid alone does not necessarily close stdin/stdout/stderr, so Gum's PTY readers can remain open and Gum never observes a fully finished command.

For a fire-and-forget launcher, do not keep the application under gum spin; detach it and redirect all inherited streams:

    setsid --fork flatpak run org.mozilla.firefox \
      </dev/null >/dev/null 2>&1
    exit 0

Then launch the script with:

    gnome-terminal --wait -- $menu_path

--wait makes the gnome-terminal client wait for the short launcher command; once setsid --fork returns and the script exits, the terminal can close while Firefox continues in its own session with no inherited terminal streams.

Why this matters specifically with Gum: gum spin starts the child with exec.CommandContext, attaches its stdin, and on a TTY wires stdout/stderr through PTYs while waiting for the process. A descendant that retains those PTY descriptors can keep the supervision path alive. See [spin.commandStart](https://github.com/charmbracelet/gum/blob/7179388031ae67d7f538d001be87d931f1cf5e28/spin/spin.go#L58-L113).

If you want a spinner only during startup, supervise a bounded readiness command, not the long-lived GUI process. If this closes the terminal while keeping Firefox alive, please mark the reply as accepted.
